[en] Blog post on Confidential Kubernetes

datosh opened the pull request:
Signed-off-by: Fabian Kammel <fk@edgeless.systems>

Hey 👋 

This blog post has been in the making for quite some time. [The initial idea was discussed 7 months ago](https://kubernetes.slack.com/archives/C019LFTGNQ3/p1654274804381029). All further discussions can also be [found in the same Slack thread](https://kubernetes.slack.com/archives/C019LFTGNQ3/p1673954867666159?thread_ts=1654274804.381029&cid=C019LFTGNQ3) posted in #sig-security & #sig-security-docs.

Over the holidays I have gathered all the bits and pieces [we've written to the initial HackMD file](https://hackmd.io/5-Muk9X_SyW2bGf75sOsUw) and made it into this frist version of the blog post.

I have also presented the status in #sig-security-docs meeting in December & Januar, and they encourged me to push this forward to GitHub.

There are still some `TODO`s left in the document where I am specifically looking for feedback/input. Feel free to provide and pointers, guidance & feedback on all other parts as well 🙂 

Please, add yourself as an author if you have contributed to the document! Sadly, HackMD is quite anonymous about who contributed what, so I am not able to mention the contributors from that information. 

CC: @mythi @fitzthum @m1ghtym0 

<!--

 Hello!

 Remember to ADD A DESCRIPTION and delete this note before submitting
 your pull request. The description should explain what will change,
 and why.

 PLEASE title the FIRST commit appropriately, so that if you squash all
 your commits into one, the combined commit message makes sense.
 For overall help on editing and submitting pull requests, visit:
  https://kubernetes.io/docs/contribute/suggesting-improvements/

 Use the default base branch, “main”, if you're documenting existing
 features in the English localization.

 If you're working on a different localization (not English), see
 https://kubernetes.io/docs/contribute/new-content/overview/#choose-which-git-branch-to-use
 for advice.

 If you're documenting a feature that will be part of a future release, see
 https://kubernetes.io/docs/contribute/new-content/new-features/ for advice.

-->

windsonsea commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+
+Zero Knowledge Proofs or Protocols are a privacy-preserving technique (PPT) that allows one party to prove facts about their data without revealing anything else about the data. ZKP can be used instead of or in addition to confidential computing to protect the privacy of the involved parties and their data. Similarly, Multi-Party Computation enables multiple parties to work together on a computat ...
+
+# Use Cases of Confidential Computing
```
```suggestion
## Use Cases of Confidential Computing
```
A topic only can have one heading 1, but this PR has four now.

windsonsea commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+**Authors:** Fabian Kammel (Edgeless Systems), (more to add...)
+
+
+# Confidential Kubernetes: Use Confidential Virtual Machines and Enclaves to improve your cluster security.
```
```suggestion
# Confidential Kubernetes: Use Confidential Virtual Machines and Enclaves to improve your cluster security.
```
This heading can be removed while the title has the same role.

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+slug: "confidential-kubernetes"
+---
+
+**Authors:** Fabian Kammel (Edgeless Systems), (more to add...)
```
Noting that this is pending a fix

/retitle [WIP] Add blog post on Confidential Kubernetes

datosh replied:
I have added @mythi & @fitzthum as additional authors and removed the TODO.

Let me know (or propose a change) to add yourself as an author!

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+
+In this blog post, we will introduce the concept of Confidential Computing (CC) to improve the security and privacy properties of any computing environment. Further, we will show how the Cloud-Native ecosystem, and Kubernetes in particular, can benefit from the new compute paradigm.
+
+## Intro
```
You don't need to have a heading to say that you're introducing the article; this can be implied. I would remove this heading or change the wording.

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+
+## Intro
+
+The [Whitepaper](https://confidentialcomputing.io/wp-content/uploads/sites/85/2022/11/CCC-A-Technical-Analysis-of-Confidential-Computing-v1.2_updated_2022-11-02.pdf) produced by the [Confidential Computing Consortium](https://confidentialcomputing.io/), a [LinuxFoundation](https://confidentialcomputing.io/wp-content/uploads/sites/85/2019/12/CCC_Overview.pdf) project, provides a great motivation f ...
```
- “Linux Foundation” (with a space)
- To hyperlink to https://confidentialcomputing.io/wp-content/uploads/sites/85/2019/12/CCC_Overview.pdf, use the link text “Defining and Enabling Confidential Computing” (you will need to reword)

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+
+The [Confidential Containers project](https://github.com/confidential-containers) enables users to run a container inside a confidential context. It provides an abstraction layer so that users do not need to interface directly with the confidential computing hardware.
+
+Confidential Containers is a [CNCF sandbox project](https://www.redhat.com/en/blog/what-confidential-containers-project).
```
This isn't quite the right link.
```suggestion
[Confidential Containers](https://www.cncf.io/projects/confidential-containers/) is a CNCF sandbox project.
```

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+
+Confidential Containers is a [CNCF sandbox project](https://www.redhat.com/en/blog/what-confidential-containers-project).
+
+> TODO: @Tobin/Mikko (or other CoCo folks) can you expand on this chapter?
```
Calling out a **TODO** here.

sftim replied:
```suggestion
CoCo offers an extension to Kubernetes clusters that are comprised of bare metal nodes
that support TEEs, to launch containerized workloads within a TEE created on the fly. The
CoCo container runtime first launches a TEE, then demonstrates its validity, followed by
verifying the container image signature and seeking any decryption keys to launch the
container.

Note the entire image handling occurs within a TEE. Coco's near term goals are to include
non-bare metal Kubernetes clusters and further separate management and tenant APIs
to restrict the Kubernetes cluster admin access to the tenant pods and the containers within.
CoCo currently supports both AMD and Intel TEE solutions. It is defining both a generic
Key Broker and Attestation service to support a end-to-end confidential flow.
```
from https://github.com/mkbhanda in https://github.com/kubernetes/website/pull/38973/files#r1084549379

datosh replied:
> Calling out a **TODO** here.

I have removed the TODO since @mkbhanda provided a great summary of CoCo. 

Still if you have any further improvements on the chapter let me know or suggest edits 🙂

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+
+## Managed Confidential Kubernetes
+
+[Azure](https://learn.microsoft.com/en-us/azure/confidential-computing/confidential-node-pool-aks) and [GCP](https://cloud.google.com/blog/products/identity-security/announcing-general-availability-of-confidential-gke-nodes) both support the use of confidential virtual machines as worker nodes for their managed Kubernetes offerings.
+
+Both services aim for better workload protection and security guarantees by enabling memory encryption for container workloads. However, they don't seek to fully isolate the cluster or workloads against the service provider or infrastructure. Specifically, they don't offer a dedicated confidential control plane or expose attestation capabilities for the confidential cluster/nodes.
+
+Azure also enables [Confidential Containers](https://github.com/confidential-containers) in their managed Kubernetes offering. They support the creation based on [Intel SGX enclaves](https://learn.microsoft.com/en-us/azure/confidential-computing/confidential-containers-enclaves) and [AMD SEV-based VMs](https://techcommunity.microsoft.com/t5/azure-confidential-computing/microsoft-introduces-previe ...
```
Can we list other providers, eg Alibaba - https://www.alibabacloud.com/help/en/container-service-for-kubernetes/latest/tee-based-confidential-computing-tee-based-confidential-computing?

We should also assert what the trust model / threat model looks like for a provider. For example, if there is no attestation capability, do you need to trust that your provider is delivering the service they claim to? Also, do we now rely on the CPU vendor more than we previously did?

sunstonesecure-robert replied:
from what I can tell Alibaba's ACK is using inclavare and occulum.  which is great! but we should probably list Inclavare and Occulum at the top level, and examples of *using* those as use cases.  Same for Azure too I think.

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+
+[Constellation is a Kubernetes engine](https://github.com/edgelesssys/constellation) that aims to provide the best possible data security. It wraps your entire Kubernetes cluster into a single confidential context that is shielded from the underlying cloud infrastructure. Everything inside is always encrypted, including at runtime in memory. It shields both the worker and control plane nodes. In  ...
+
+> TODO: Please add any interesting open-source projects in the space we should highlight!
```
Calling out a **TODO** here.

datosh replied:
Resolved

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+
+## Constellation
+
+[Constellation is a Kubernetes engine](https://github.com/edgelesssys/constellation) that aims to provide the best possible data security. It wraps your entire Kubernetes cluster into a single confidential context that is shielded from the underlying cloud infrastructure. Everything inside is always encrypted, including at runtime in memory. It shields both the worker and control plane nodes. In  ...
```
```suggestion
[Constellation](https://github.com/edgelesssys/constellation) is a Kubernetes engine that aims to provide the best possible data security. Constellaton wraps your entire Kubernetes cluster into a single confidential context that is shielded from the underlying cloud infrastructure. Everything inside is always encrypted, including at runtime in memory. It shields both the worker and control plane nodes. In addition, it already integrates with popular CNCF software such as Cilium for secure networking and provides extended CSI drivers to write data securely.
```

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+If you are currently working on a high-security product that struggles to run in the public cloud due to legal requirements or are looking to bring the privacy and security of your cloud-native project to the next level: Reach out to all the great projects we have highlighted! Everyone is keen to improve security of our ecosystem and you can play a vital role in that journey.
+
+* [Confidential Containers](https://github.com/confidential-containers)
+* [Constellation: Always encrypted K8s](https://github.com/edgelesssys/constellation)
```
```suggestion
* [Constellation: Always Encrypted Kubernetes](https://github.com/edgelesssys/constellation)
```

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+* [Confidential Containers](https://github.com/confidential-containers)
+* [Constellation: Always encrypted K8s](https://github.com/edgelesssys/constellation)
+
+> TODO: Do we have a full list of CNCF-related CC projects? We should link to their GitHub to encourage adoption and contributions.
```
Calling out a **TODO** here.

sunstonesecure-robert replied:
maybe link to: https://confidentialcomputing.io/projects/

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+
+Intel SGX-based TEEs are hard to benchmark as [shown](https://arxiv.org/pdf/2205.06415.pdf) [by](https://www.ibr.cs.tu-bs.de/users/mahhouk/papers/eurosec2021.pdf) [different papers](https://dl.acm.org/doi/fullHtml/10.1145/3533737.3535098). Since enclaves are process isolated, the operating system's libraries are not available and additional SDKs are required to compile programs for SGX. The chose ...
+
+Confidential virtual machines based on AMD SEV-SNP require no changes to the executed program and operating system and are a lot easier to benchmark. A [benchmark from Azure and AMD shows](https://community.amd.com/t5/business/microsoft-azure-confidential-computing-powered-by-3rd-gen-epyc/ba-p/497796) that SEV-SNP VM overhead is <10%, sometimes as low as 2%.
```
```suggestion
Confidential virtual machines based on AMD SEV-SNP require no changes to the executed program and operating system and are a lot easier to benchmark. A [benchmark from Azure and AMD](https://community.amd.com/t5/business/microsoft-azure-confidential-computing-powered-by-3rd-gen-epyc/ba-p/497796) shows that SEV-SNP VM overhead is <10%, sometimes as low as 2%.
```

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+
+The 3rd Generation Xeon CPUs (aka Ice Lake Server - "ICX") did switch to using a technology called [Total Memory Encryption - Multi-Key (TME-MK)](https://www.intel.com/content/www/us/en/developer/articles/news/runtime-encryption-of-memory-with-intel-tme-mk.html) that uses AES-XTS, moving away from the [Memory Encryption Engine](https://eprint.iacr.org/2016/204.pdf) that the consumer and Xeon E CP ...
+
+A complete list of supported platforms is available on [ark.intel.com](https://ark.intel.com/content/www/us/en/ark/search/featurefilter.html?productType=873&2_SoftwareGuardExtensions=Yes%20with%20both%20Intel%C2%AE%20SPS%20and%20Intel%C2%AE%20ME).
```
```suggestion
A [list of supported platforms](https://ark.intel.com/content/www/us/en/ark/search/featurefilter.html?productType=873&2_SoftwareGuardExtensions=Yes%20with%20both%20Intel%C2%AE%20SPS%20and%20Intel%C2%AE%20ME) is available from Intel.
```

fitzthum commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,152 @@
+
+The **Confidentiality** property ensures that information cannot be viewed while it is in use in the TEE. This provides us with the highly desired feature to secure **data in use**. Depending on the specific TEE used, both code and data may be protected from outside viewers.
+
+Confidentiality is a great feature, but an attacker would still be able to manipulate or inject arbitrary code and data for the TEE to execute and therefore easily leak critical information. **Integrity** allows a TEE owner to check that neither code nor data has been tempered before running critical computations.
```
`Integrity allows a TEE owner to check that neither code nor data has been tempered before running critical computations.` --> `Integrity guarantees a TEE owner that neither code nor data can be tampered with while running critical computations.`

The current wording, particularly where we say that it allows an owner to check the code before a computation, sounds a bit more like attestation to me.

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+
+The enclave memory cannot be read or written from outside the enclave, regardless of the current privilege level and CPU mode. The only way to call an enclave function is through a new instruction that performs several protection checks. Its memory is encrypted. Tapping the memory or connecting the DRAM modules to another system will yield only encrypted data. The memory encryption key randomly c ...
+
+The 3rd Generation Xeon CPUs (aka Ice Lake Server - "ICX") did switch to using a technology called [Total Memory Encryption - Multi-Key (TME-MK)](https://www.intel.com/content/www/us/en/developer/articles/news/runtime-encryption-of-memory-with-intel-tme-mk.html) that uses AES-XTS, moving away from the [Memory Encryption Engine](https://eprint.iacr.org/2016/204.pdf) that the consumer and Xeon E CP ...
```
```suggestion
The 3rd generation Xeon CPUs (aka Ice Lake Server - "ICX") did switch to using a technology called [Total Memory Encryption - Multi-Key](https://www.intel.com/content/www/us/en/developer/articles/news/runtime-encryption-of-memory-with-intel-tme-mk.html)  (TME-MK) that uses AES-XTS, moving away from the [Memory Encryption Engine](https://eprint.iacr.org/2016/204.pdf) that the consumer and Xeon E CPUs used. This increased the possible [enclave page cache](https://sgx101.gitbook.io/sgx101/sgx-bootstrap/enclave#enclave-page-cache-epc) size (up to 512GB/CPU) as well as providing a big increase in performance.

You can find more information about Intel SGX on multi-socket platforms in the whitepaper [Supporting Intel® SGX on Multi-Socket Platforms](https://www.intel.com/content/dam/www/public/us/en/documents/white-papers/supporting-intel-sgx-on-mulit-socket-platforms.pdf).
```

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+[Intel's Software Guard Extensions](https://www.intel.com/content/www/us/en/developer/tools/software-guard-extensions/overview.html) is available since 2015 [introduced with the Skylake architecture](https://ark.intel.com/content/www/us/en/ark/products/88172/intel-xeon-processor-e31220-v5-8m-cache-3-00-ghz.html).
+
+SGX is a new instruction set that enables users to create a protected and isolated process called an *enclave*.
+It provides a reverse sandbox that protects enclaves from the OS, BIOS, firmware, and any other privileged execution context.
```
```suggestion
It provides a reverse sandbox that protects enclaves from the operating system, firmware, and any other privileged execution context.
```
A BIOS is a kind of firmware.

fitzthum commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,152 @@
+
+*Availability* is the third basic property often [discussed in information security](https://en.wikipedia.org/wiki/Information_security). This property is outside the scope of most TEEs. Usually, they can be controlled (shutdown, restarted, ...) by some higher-level concept. This could be the CPU itself, a hypervisor, or a kernel. This is to preserve the availability of the overall system, and no ...
+
+Confidentiality and Integrity by themselves would only be helpful in some cases. For example, consider a TEE running in a remote cloud. How would you know that the TEE is genuine and running your intended software? It could be an imposter stealing your data as soon as you send it over. This fundamental problem is addressed by **Attestability**. Remote attestation allows us to verify the identity, ...
```
Is the word remote appropriate in the last sentence?

datosh replied:
Good point, I've added a sentence to clarify.

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+
+### Intel SGX
+
+[Intel's Software Guard Extensions](https://www.intel.com/content/www/us/en/developer/tools/software-guard-extensions/overview.html) is available since 2015 [introduced with the Skylake architecture](https://ark.intel.com/content/www/us/en/ark/products/88172/intel-xeon-processor-e31220-v5-8m-cache-3-00-ghz.html).
```
```suggestion
Intel's [Software Guard Extensions](https://www.intel.com/content/www/us/en/developer/tools/software-guard-extensions/overview.html) are available since 2015, and were introduced with the Skylake architecture.
```
I removed the link because https://ark.intel.com/content/www/us/en/ark/products/88172/intel-xeon-processor-e31220-v5-8m-cache-3-00-ghz.html is not a page about Skylake; it is a page about a particular Intel SKU / product.

fitzthum commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,152 @@
+
+Confidentiality and Integrity by themselves would only be helpful in some cases. For example, consider a TEE running in a remote cloud. How would you know that the TEE is genuine and running your intended software? It could be an imposter stealing your data as soon as you send it over. This fundamental problem is addressed by **Attestability**. Remote attestation allows us to verify the identity, ...
+
+TEEs can hold and process information that needs to persist even beyond their lifespan. That could mean across restarts, different versions, or platform migrations. Therefore **Recoverability** is an important feature. Data and the state of a TEE need to be sealed before they are written to persistent storage to maintain confidentiality and integrity guarantees. The access to such sealed data nee ...
```
I think we could potentially leave this paragraph out as well as the section on migration.

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+
+SEV is continually upgraded with new features and improvements with each new generation. The Linux community makes these features available as part of the KVM hypervisor, host, and guest kernel. The first SEV features were discussed and implemented in [2016](https://www.usenix.org/conference/usenixsecurity16/technical-sessions/presentation/kaplan). After that, the SEV-ES feature set was implement ...
+
+Confidential VMs based on AMD SEV-SNP are [available in Microsoft Azure since July 2022](https://azure.microsoft.com/en-us/updates/azureconfidentialvm/). Similarly, [Google Cloud Platform (GCP)](https://cloud.google.com/compute/confidential-vm/docs/about-cvm) offers confidential VMs based on AMD SEV-ES.
```
```suggestion
[Confidential VMs based on AMD SEV-SNP](https://azure.microsoft.com/en-us/updates/azureconfidentialvm/)) are available in Microsoft Azure since July 2022. Similarly, Google Cloud Platform offers [confidential VMs based on AMD SEV-ES](https://cloud.google.com/compute/confidential-vm/docs/about-cvm).
```

fitzthum commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,152 @@
+
+AMD's [Secure Encrypted Virtualization (SEV)](https://developer.amd.com/sev/) technologies are a set of features to enhance the security of virtual machines on AMD's server CPUs. SEV transparently encrypts the memory of each VM with a unique key.
+SEV can also calculate a signature of the memory contents, which can be sent to the VM's owner as an attestation that the memory was encrypted correctly by the firmware.
+The second generation of SEV called [Encrypted State (SEV-ES)](https://www.amd.com/system/files/TechDocs/Protecting%20VM%20Register%20State%20with%20SEV-ES.pdf) provides additional protection from the hypervisor by encrypting all CPU register contents when a VM stops running.
```
I think that technically, SEV-ES is a feature, rather than a generation of SEV. It might be more accurate to say that the second generation of AMD EPYC processors (Rome) support a feature called SEV-ES.

fitzthum commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,152 @@
+AMD's [Secure Encrypted Virtualization (SEV)](https://developer.amd.com/sev/) technologies are a set of features to enhance the security of virtual machines on AMD's server CPUs. SEV transparently encrypts the memory of each VM with a unique key.
+SEV can also calculate a signature of the memory contents, which can be sent to the VM's owner as an attestation that the memory was encrypted correctly by the firmware.
+The second generation of SEV called [Encrypted State (SEV-ES)](https://www.amd.com/system/files/TechDocs/Protecting%20VM%20Register%20State%20with%20SEV-ES.pdf) provides additional protection from the hypervisor by encrypting all CPU register contents when a VM stops running.
+The third generation of SEV called [Secure Nested Paging (SNP)](https://www.amd.com/system/files/TechDocs/SEV-SNP-strengthening-vm-isolation-with-integrity-protection-and-more.pdf) is designed to prevent software-based integrity attacks and reduce the risk associated with
```
Same as above.

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,155 @@
+a private (encrypted) memory page, it must always read the value it last wrote.
+Additionally, by allowing the guest to obtain remote attestation statements dynamically, SNP enhances the remote attestation capabilities of SEV.
+
+SEV is continually upgraded with new features and improvements with each new generation. The Linux community makes these features available as part of the KVM hypervisor, host, and guest kernel. The first SEV features were discussed and implemented in [2016](https://www.usenix.org/conference/usenixsecurity16/technical-sessions/presentation/kaplan). After that, the SEV-ES feature set was implement ...
```
Try this:
```suggestion
AMD SEV is continually upgraded with new features and improvements with each new generation. The Linux community makes these features available as part of the KVM hypervisor, and for host and guest kernels. The first SEV features were discussed and implemented in 2016 - see [AMD x86 Memory Encryption Technologies](https://www.usenix.org/conference/usenixsecurity16/technical-sessions/presentation/kaplan) from the 2016 Usenix Security Symposium.

After that, the SEV-ES feature set was [implemented](https://www.phoronix.com/news/AMD-SEV-ES-Linux-2020-Patches) for Linux in 2020. Recently, an updated feature set was [released](https://www.phoronix.com/news/AMD-SEV-SNP-Arrives-Linux-5.19) as part of Linux 5.19 in July 2022, which enabled support for SEV-SNP.
```

fitzthum commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,152 @@
+a private (encrypted) memory page, it must always read the value it last wrote.
+Additionally, by allowing the guest to obtain remote attestation statements dynamically, SNP enhances the remote attestation capabilities of SEV.
+
+SEV is continually upgraded with new features and improvements with each new generation. The Linux community makes these features available as part of the KVM hypervisor, host, and guest kernel. The first SEV features were discussed and implemented in [2016](https://www.usenix.org/conference/usenixsecurity16/technical-sessions/presentation/kaplan). After that, the SEV-ES feature set was implement ...
```
It might be better to say that SEV has been implemented incrementally rather than say that it is continually upgraded, because that implies that it will continue to be upgraded, which we can't really confirm.

fitzthum commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,152 @@
+
+### Intel SGX
+
+[Intel's Software Guard Extensions](https://www.intel.com/content/www/us/en/developer/tools/software-guard-extensions/overview.html) is available since 2015 [introduced with the Skylake architecture](https://ark.intel.com/content/www/us/en/ark/products/88172/intel-xeon-processor-e31220-v5-8m-cache-3-00-ghz.html).
```
`is available` -> `has been available`.

fitzthum commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,152 @@
+[Intel's Software Guard Extensions](https://www.intel.com/content/www/us/en/developer/tools/software-guard-extensions/overview.html) is available since 2015 [introduced with the Skylake architecture](https://ark.intel.com/content/www/us/en/ark/products/88172/intel-xeon-processor-e31220-v5-8m-cache-3-00-ghz.html).
+
+SGX is a new instruction set that enables users to create a protected and isolated process called an *enclave*.
+It provides a reverse sandbox that protects enclaves from the OS, BIOS, firmware, and any other privileged execution context.
```
Reverse sandbox is an interesting term.  It is only used in the SGX section, but it seems like it would apply to enclaves more generally? It implies that there  is little protection of the host from the guest, is that true for SGX?

datosh replied:
I have not written this particular sentence, can someone chime in @mythi @m1ghtym0 ?

sftim reviewed (commented):
Thanks for this draft. Here's some feedback that will help you get this ready for approval:

- Please write all headings within the page in _sentence case_ (usually, first letter capitalized).
- Wrap the Markdown source at around 100 characters; this helps contributors that work with the article text after merge, such as localization teams.
- Is there room to mention IBM Secure Execution for Linux?

fitzthum commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,152 @@
+
+> TODO: Should we mention attacks? If we do, we should also mention attacks on all technologies. Do we have a stance / recommendation here or do we provide the information and let the reader decide?
+
+SGX is available on [Azure](https://azure.microsoft.com/de-de/updates/intel-sgx-based-confidential-computing-vms-now-available-on-azure-dedicated-hosts/), [Alibaba Cloud](https://azure.microsoft.com/de-de/updates/intel-sgx-based-confidential-computing-vms-now-available-on-azure-dedicated-hosts/), [IBM](https://cloud.ibm.com/docs/bare-metal?topic=bare-metal-bm-server-provision-sgx), and many more.
```
It might be worth noting that SGX originally had a limited memory capacity, but that this has been remedied in more recent release. Also, SGX has typically required modification  of a workload or use of a library OS.

It is probably beyond the scope to compare overhead and frequency of context switches, but might be something to allude to.

fitzthum commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,152 @@
+
+Where Intel SGX aims to protect the context of a single process, [Intel's Trusted Domain Extensions](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-trust-domain-extensions.html) protects a full virtual machine and is therefore most closely comparable to AMD SEV.
+
+As with SEV-SNP support for TDX was [merged in Linux Kernel 5.19](https://www.phoronix.com/news/Intel-TDX-For-Linux-5.19), but hardware support will land with [Saphire Rapids](https://en.wikipedia.org/wiki/Sapphire_Rapids) in early 2023.
```
This is only guest support, right?

m1ghtym0 replied:
Yes, the host support is still being [discussed](https://lore.kernel.org/linux-mm/755f94aa-a0cc-b7d6-ce8a-a81ff4f598da@intel.com/T/).

sftim commented:
/hold

Publication date is not yet assigned (by blog team)

fitzthum commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,152 @@
+
+## Overhead Analysis
+
+The benefits that confidential computing technologies provide via strong isolation and enhanced security to customer data and workloads are not for free. Quantifying this impact isn't easy and depends on many factors: The TEE technology, the benchmark, the metrics, and the type of workload have a huge impact on the induced performance overhead.
```
I think we could maybe get rid of this section. We don't really offer any meaningful conclusions. Instead maybe we can briefly mention performance in the section above.

fitzthum commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,152 @@
+
+Although there is a performance overhead, it should be low enough to enable real-world workloads to run in these protected environments and improve the security and privacy of our data.
+
+## Confidential Computing compared to FHE, ZKP, and MPC
```
Don't think this is necessary.

datosh replied:
I would also like to keep it. Lot's of folks have heard about the concepts (mostly FHE). This allows us to highlight that CC (in general) is more user friendly and has additional properties.

fitzthum commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,152 @@
+
+Confidential Containers is a [CNCF sandbox project](https://www.redhat.com/en/blog/what-confidential-containers-project).
+
+> TODO: @Tobin/Mikko (or other CoCo folks) can you expand on this chapter?
```
Yes

datosh replied:
Awesome - thanks so much for providing this @mkbhanda!

fitzthum reviewed (commented):
Lots of good stuff here. 

I think in general we spend a lot of time talking about details of CC before we get into any cloud native stuff. Is this intended as a general purpose guide to CC or are we focusing more on the cloud native integration? If it's the latter, I think we should try to move that forward in the post and try to offer a more substantial structure to the cloud native part. For example, can we compare TCB, deployment models, etc.

datosh commented:
Thanks for all the great feedback @sftim !

> * Please write all headings within the page in _sentence case_ (usually, first letter capitalized).

Done. I have kept the capitalization on terms like "Confidential Computing" and "Confidential Containers". I hope this is in line with the conventions.

> * Wrap the Markdown source at around 100 characters; this helps contributors that work with the article text after merge, such as localization teams.

Done. Some lines are a lot longer due to long links. Is it okay or should I move links to their own section of the document and use shorter references in text?

> * Is there room to mention IBM Secure Execution for Linux?

Yes! Would you be open to supply this section?

igordcard commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,312 @@
+[Confidential Computing Consortium](https://confidentialcomputing.io/) is a Linux Foundation
+that already worked on
+[Defining and Enabling Confidential Computing](https://confidentialcomputing.io/wp-content/uploads/sites/85/2019/12/CCC_Overview.pdf).
+In the [Whitepaper](https://confidentialcomputing.io/wp-content/uploads/sites/85/2022/11/CCC-A-Technical-Analysis-of-Confidential-Computing-v1.2_updated_2022-11-02.pdf)
```
Whitepaper link seems to be dead.. link to v1.3 works:
https://confidentialcomputing.io/wp-content/uploads/sites/85/2023/01/CCC-A-Technical-Analysis-of-Confidential-Computing-v1.3_Updated_November_2022.pdf

mkbhanda reviewed (commented):
Good work on a comprehensive compilation! Suggesting minor English edits and some feedback on security gaps references.

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,312 @@
+**Authors:** Fabian Kammel (Edgeless Systems), (more to add...)
+
+In this blog post, we will introduce the concept of Confidential Computing (CC) to improve
+the security and privacy properties of any computing environment. Further, we will show how
```
“we will” is fine, honestly

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,312 @@
+
+*Availability* is the third basic property often discussed in the context of information
+security. This property is outside the scope of most TEEs. Usually, they can be controlled
+(shut down, restarted, …) by some higher-level concept. This could be the CPU itself, a
```
Maybe “higher level abstraction”.

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,312 @@
+from the hardware itself. This feature can also be made available to clients outside of the
+confidential computing hardware in the form of remote attestation.
+
+TEEs can hold and process information that needs to persist even beyond their lifespan. That
```
```suggestion
TEEs can hold and process information that predates or outlives the trusted environment. That
```

How about this?

mkbhanda replied:
looks good, it highlights that the workload is the thing that lives on, the TEE is a vessel.

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,312 @@
+
+TEEs can hold and process information that needs to persist even beyond their lifespan. That
+could mean across restarts, different versions, or platform migrations. Therefore **Recoverability**
+is an important feature. Data and the state of a TEE need to be sealed before they are written
```
Nope: “Recoverability is an important feature” is idiomatic English.

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,312 @@
+
+This does not have to limit the flexibility of the overall system. The
+[migration agent (MA) of AMD SEV-SNP](https://www.amd.com/system/files/TechDocs/SEV-SNP-strengthening-vm-isolation-with-integrity-protection-and-more.pdf)
+allows users to migrate a confidential virtual machine to a different host system. All
```
You didn't specify a replacement?

datosh replied:
I think the suggestion was to merge the sentence.

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,312 @@
+
+Intel's
+[Software Guard Extensions](https://www.intel.com/content/www/us/en/developer/tools/software-guard-extensions/overview.html)
+has been available since 2015, and were introduced with the Skylake architecture.
```
“has” is idiomatic English, I'm afraid

datosh replied:
```suggestion
has been available since 2015 and were introduced with the Skylake architecture.
```

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,312 @@
+is available from Intel.
+
+However, there are still several security limitations faced by SGX. Such as Cache
+timing attacks, physical attacks, untrusted I/O, or malicious microcode patching.
```
I'm happy with “or” in this context.

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,312 @@
+
+Where Intel SGX aims to protect the context of a single process,
+[Intel's Trusted Domain Extensions](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-trust-domain-extensions.html)
+protects a full virtual machine and is therefore most closely comparable to AMD SEV.
```
I prefer
```suggestion
protect a full virtual machine and are therefore most closely comparable to AMD SEV.
```

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,312 @@
+
+Both services aim for better workload protection and security guarantees by enabling memory encryption
+for container workloads. However, they don't seek to fully isolate the cluster or workloads against
+the service provider or infrastructure. Specifically, they don't offer a dedicated confidential control
```
This changes the meaning. Remember, not all workloads are hosted with a supplier. I like the original.

datosh replied:
I will keep the original.

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,312 @@
+## Where are we today? Vendors, limitations, and FOSS landscape
+
+As we have seen in the previous sections Confidential Computing is a powerful new concept
+to improve security, but we are still in the (early) adoption phase. New products are
```
I'm happy with your original text here @datosh

datosh replied:
I changed `new properties` to `unique properties` to remove the repeated usage of `new` in the same sentence.

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,312 @@
+starting to emerge to take advantage of the new properties.
+
+Google and Microsoft are the first major cloud providers to have confidential offerings.
+Still, these offerings are limited to compute and end-to-end solutions for confidential
```
I'm happy with your original text here @datosh

datosh replied:
Accepted "while", it highlights the gap in actual service offerings clearer.

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,312 @@
+
+### Confidential Containers
+
+The [Confidential Containers project](https://github.com/confidential-containers) enables users
```
How about, then:
```suggestion
The [Confidential Containers](https://github.com/confidential-containers) (CoCo) project enables users
```

datosh commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,312 @@
+A [list of supported platforms](https://ark.intel.com/content/www/us/en/ark/search/featurefilter.html?productType=873&2_SoftwareGuardExtensions=Yes%20with%20both%20Intel%C2%AE%20SPS%20and%20Intel%C2%AE%20ME)
+is available from Intel.
+
+However, there are still several security limitations faced by SGX. Such as Cache
```
I have rewritten this section and removed the links to specific attacks.

datosh commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,312 @@
+Among others [Prime+Probe](https://gruss.cc/files/malware_guard_extension.pdf) and
+[SGAxe](https://sgaxe.com/files/SGAxe.pdf).
+
+> TODO: Should we mention attacks? If we do, we should also mention attacks on all technologies. Do we have a stance / recommendation here or do we provide the information and let the reader decide?
```
I have removed links to attack. A full analysis of attacks for all technologies is out of scope.

datosh commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,312 @@
+confidential computing hardware in the form of remote attestation.
+
+TEEs can hold and process information that needs to persist even beyond their lifespan. That
+could mean across restarts, different versions, or platform migrations. Therefore **Recoverability**
```
I would like to keep as is, since this list is not meant to be exhaustive, but only hint at different concepts. Pause and restart are similar in that regard.

datosh commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,312 @@
+to persistent storage to maintain confidentiality and integrity guarantees. The access to such
+sealed data needs to be well-defined. In most cases, the unsealing is bound to a TEE's identity.
+Hence, making sure the recovery can only happen in the same confidential context.
+
```
I am not sure what you are suggesting here. Is this a feature you would like to see supported in future versions of SNP/TDX?

datosh commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,312 @@
+
+If you are currently working on a high-security product that struggles to run in the
+public cloud due to legal requirements or are looking to bring the privacy and security
+of your cloud-native project to the next level: Reach out to all the great projects
```
Technical is good, but lets have some good vibes to get folks excited to contribute 🙂

datosh commented:
Thanks for all the feedback! I have responded to all comments and included most suggestions.

Most of the feedback I received was mainly on structure and wording, i.e., that content we have already produced. Is anyone still writing or is planning to write more content to be added to this blog post?
* @sftim would you like to add something regarding IBM Secure Execution for Linux? 
* @mythi @fitzthum would you like to extend/revise the chapter on CoCo or otherwise provide more CC content? 

Also who would like to be added to the authors section? Please let me know or add yourself!

@sftim what are the next steps to get this post finalized and closer to publication? 
The two open TODOs are an invitation for contribution and can be removed anytime.

igordcard commented:
@datosh I can add a couple tweaks from my latest pass on it, and another "secure project to emerge" section for Gramine (https://gramineproject.io/). I can go ahead and push a commit on top if that's the preferred method.

datosh commented:
> @datosh I can add a couple tweaks from my latest pass on it, and another "secure project to emerge" section for Gramine (https://gramineproject.io/). I can go ahead and push a commit on top if that's the preferred method.

Sound great, @igordcard! Yes, feel free to push your changes directly.

mythi reviewed (commented):
1st round of comments. I'm planning to work on adding something about TCB and sync with @igordcard on some of the topics he mentioned.

Having written a couple of kubernetes.io blogs and based on the comments I've received earlier, I think we may still need to add more glue to the beginning to justify why Kubernetes developers or end users should read through all the HW intros till the end: for example, what industries and why they need CC in clouds and how kubernetes can help to satisfy that requirement.

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,315 @@
+protect a full virtual machine and are therefore most closely comparable to AMD SEV.
+
+As with SEV-SNP, guest support for TDX was [merged in Linux Kernel 5.19](https://www.phoronix.com/news/Intel-TDX-For-Linux-5.19),
+but hardware support will land with [Saphire Rapids](https://en.wikipedia.org/wiki/Sapphire_Rapids) in early 2023.
```
Avoid hard statements about the future unless _the Kubernetes project_ is committing to making these come true.

sftim commented:
> @sftim would you like to add something regarding IBM Secure Execution for Linux

How about:
_IBM Secure Execution for Linux lets you run your Kubernetes cluster's nodes as KVM guests within a trusted execution environment on IBM Z series hardware. You can use this hardware-enhanced virtual machine isolation to provide strong isolation between tenants in a cluster, with hardware attestation about the (virtual) node's integrity._

and let people read between the lines that you can't get Pod-level secure execution with this solution.

Mention this at around the same place as ARM TrustZone, perhaps?

datosh commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,315 @@
+the service provider or infrastructure. Specifically, they don't offer a dedicated confidential control
+plane or expose attestation capabilities for the confidential cluster/nodes.
+
+Azure also enables [Confidential Containers](https://github.com/confidential-containers)
+in their managed Kubernetes offering. They support the creation based on
+[Intel SGX enclaves](https://learn.microsoft.com/en-us/azure/confidential-computing/confidential-containers-enclaves)
```
I have replaced the link with: https://learn.microsoft.com/en-us/azure/confidential-computing/confidential-nodes-aks-overview

Do you think this is more fitting? I have not used any confidential computing features of the AKS offering myself.

datosh commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,315 @@
+
+SGX is available on
+[Azure](https://azure.microsoft.com/de-de/updates/intel-sgx-based-confidential-computing-vms-now-available-on-azure-dedicated-hosts/),
+[Alibaba Cloud](https://azure.microsoft.com/de-de/updates/intel-sgx-based-confidential-computing-vms-now-available-on-azure-dedicated-hosts/),
```
fixed the link

datosh commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,315 @@
+about SGX on multi-socket platforms can be found in the
+[Whitepaper](https://www.intel.com/content/dam/www/public/us/en/documents/white-papers/supporting-intel-sgx-on-mulit-socket-platforms.pdf)
+
+A [list of supported platforms](https://ark.intel.com/content/www/us/en/ark/search/featurefilter.html?productType=873&2_SoftwareGuardExtensions=Yes%20with%20both%20Intel%C2%AE%20SPS%20and%20Intel%C2%AE%20ME)
+is available from Intel.
```
Yes, the filter was too restrictive. I fixed it.

datosh commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,315 @@
+A [list of supported platforms](https://ark.intel.com/content/www/us/en/ark/search/featurefilter.html?productType=873&2_SoftwareGuardExtensions=Yes%20with%20both%20Intel%C2%AE%20SPS%20and%20Intel%C2%AE%20ME)
+is available from Intel.
+
+However, there are security limitations, such as cache timing attacks, physical attacks,
+untrusted I/O, or malicious microcode patching. Each project has different security requirements
+and needs to decide, if the added security of SGX and its limitations, are sufficient.
+
```
This is the replacement. Instead of citing specific attacks we list the classes of attacks so users can use that as a starting point for more research on the topic. Do you think this section is a good starting point for that?

sftim commented:
I'm getting the sense that this might be ready for blog team review - how does that sound?

datosh commented:
> I'm getting the sense that this might be ready for blog team review - how does that sound?

Sounds great! As far as I understand it @mythi & @igordcard are working on a few more sections. Can we do that in parallel?

fitzthum commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,324 @@
+
+### Confidential Containers
+
+The [Confidential Containers](https://github.com/confidential-containers) (CoCo) project enables users
+to process and run each container of a pod inside a confidential context.
+[Confidential Containers](https://www.cncf.io/projects/confidential-containers/) is a CNCF sandbox project.
```
```suggestion
[Confidential Containers](https://github.com/confidential-containers) (CoCo) is a CNCF sandbox project that isolates Kubernetes pods inside of confidential virtual machines.
```

datosh replied:
added

fitzthum commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,324 @@
+to process and run each container of a pod inside a confidential context.
+[Confidential Containers](https://www.cncf.io/projects/confidential-containers/) is a CNCF sandbox project.
+
+CoCo offers an extension to Kubernetes clusters that are comprised of
+nodes that support TEEs, to launch containerized workloads within a TEE created
+on the fly. With CoCo, the entire lifecycle of a pod, including the image handling occurs
+within a TEE leaving the cluster node outside of the TCB: for each pod, a TEE is launched, followed by the verification of the container images' signatures before pull and seeking any
+decryption keys to launch the container. The image policy and layer decryption keys a obtained from a trusted *key broker service* that involves a TEE remote attestation as part of the process. Coco's near term goals are to move to CSPs' Kubernetes
+clusters and further separate management and tenant APIs to restrict the
+Kubernetes cluster admin access to the tenant pods and the containers within.
+CoCo currently supports both AMD and Intel TEE solutions. It is defining both
+a generic Key Broker and Attestation service to support an end-to-end confidential
+flow.
```
```suggestion
CoCo can be installed on a Kubernetes cluster with an operator.
The operator will create a set of runtime classes that can be used to deploy
pods inside an enclave on several different platforms, including
AMD SEV, Intel TDX, Secure Execution for IBM Z, and Intel SGX.

CoCo is typically used with signed and/or encrypted container images
which are pulled, verified, and decrypted inside the enclave.
Secrets, such as image decryption keys, are conditionally provisioned
to the enclave by a trusted Key Broker Service that validates the
hardware evidence of the TEE prior to releasing any sensitive information.

CoCo has several deployment models. Since the Kubernetes control plane
is outside the TCB, CoCo is suitable for managed environments. CoCo can
be run in virtual environments that don't support nesting with the help of an
API adaptor that starts pod VMs in the cloud. CoCo can also be run on
bare metal, providing strong isolation even in multi-tenant environments.
```

fitzthum reviewed (commented):
My suggestions for the CoCo section.

fitzthum commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,324 @@
+
+The second generation of SEV, known as
+[Encrypted State](https://www.amd.com/system/files/TechDocs/Protecting%20VM%20Register%20State%20with%20SEV-ES.pdf)
+or SEV-ES provides additional protection from the hypervisor by encrypting all CPU register contents when a VM stops running.
```
`when a VM stops running` is a little bit ambiguous. To be clear, the registers are protected every time the guest stops executing i.e. a context switch.

fitzthum commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,324 @@
+After that, the SEV-ES feature set was implemented in
+[2020](https://www.phoronix.com/news/AMD-SEV-ES-Linux-2020-Patches). Recently, the latest feature set
+was [merged in Linux 5.19 in July 2022](https://www.phoronix.com/news/AMD-SEV-SNP-Arrives-Linux-5.19)
+and enabled support for SEV-SNP.
```
This is only the guest support. The host support is still pending. Honestly I don't think these sections on the timeline of the Linux implementation are necessary. They are also a bit redundant with the previous description of each feature.

datosh replied:
shortened this section a bit

fitzthum commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,324 @@
+are a set of features to enhance the security of virtual machines on AMD's server CPUs. SEV
+transparently encrypts the memory of each VM with a unique key. SEV can also calculate a
+signature of the memory contents, which can be sent to the VM's owner as an attestation that
+the memory was encrypted correctly by the firmware.
```
The attestation is about more than ensuring that the firmware correctly encrypted the initial guest memory. I think the main idea is that it guarantees to the guest owner that the initial guest memory is what the expect.

mythi commented:
> > I'm getting the sense that this might be ready for blog team review - how does that sound?
> 
> Sounds great! As far as I understand it @mythi & @igordcard are working on a few more sections. Can we do that in parallel?

I've submitted a PR to @datosh 's repo with my proposed updates and additions.

datosh commented:
Thanks for the contributions! I have added the suggestions from @mythi & @fitzthum and added them as authors.
If anyone else wants to be mentioned as author, let me know or propose the change! 

Let's move this PR forward. Is there anything I need to do to get approval from review team @sftim ?

sftim commented:
Hi @datosh 

Can you rebase this against latest _main_? Also, please consider squashing this down to a smaller number of commits whilst you do that.
We should be able to get it reviewed.

/hold
pending assignment of a publication date

sftim commented:
Hi @datosh 

I suggest we target this for late May 2023, as I'm worried that the article won't be ready for publication ahead of the upcoming Kubernetes release and associated publicity. How does that sound?

datosh commented:
I have cleaned up the PR. Could you pass this along to review & get this scheduled, @sftim ?

mythi commented:
> I have cleaned up the PR. Could you pass this along to review & get this scheduled, @sftim ?

@datosh  I submitted a PR to your branch to add a few updates.

m1ghtym0 commented:
> > I have cleaned up the PR. Could you pass this along to review & get this scheduled, @sftim ?
> 
> @datosh I submitted a PR to your branch to add a few updates.

Your changes LGTM @mythi, but only @datosh can merge them.
What would be the new target date @sftim? Since we as of today missed the end of May;-)

sftim commented:
@datosh, please rebase this PR against _main_ which should fix the previewing.

For the date:
The blog team is quite small and we're not putting as much effort into growing it as we might. You can suggest a future weekday for publication, far enough away that you think this would already have been reviewed and merged.

sftim commented:
Once this is ready for review, please remove the [WIP] marker.

datosh commented:
> Once this is ready for review, please remove the [WIP] marker.

Done!
I have added @mythi changes and rebased to main. Removed the [WIP] marker.
Please choose a publication date @m1ghtym0.
Please push this to reivew @sftim

datosh commented:
@sftim does July 6th work as a publishing date? Any input on the red pipeline (I rebased main as suggested)?

sftim commented:
Here's the deploy preview: https://deploy-preview-38973--kubernetes-io-main-staging.netlify.app/blog/2023/01/17/confidential-kubernetes/

datosh commented:
Glad the rebase resolved this. 

Please let me know about two things @sftim 
1. Does July 6th work as a release date?
2. What is left to do to get this scheduled for publishing?

sftim commented:
We're short of folks willing to review blog articles. I'm going to see if I can encourage more contributors to take part in that work.

dipesh-rawat commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,345 @@
+the Cloud-Native ecosystem, and Kubernetes in particular, can benefit from the new compute paradigm.
+
+Confidential Computing is not a new concept in the cloud-native world. The
+[Confidential Computing Consortium](https://confidentialcomputing.io/) (CCC) is a Linux Foundation
```
```suggestion
[Confidential Computing Consortium](https://confidentialcomputing.io/) (CCC) is a project community in the Linux Foundation
```

dipesh-rawat commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,345 @@
+is that they are very costly. A managed CloudHSM from AWS costs
+[around $1.50 / hour](https://aws.amazon.com/cloudhsm/pricing/) or ~$13,500 / year.
+
+In recent years a new kind of TEE has gained popularity. Technologies like
```
```suggestion
In recent years, a new kind of TEE has gained popularity. Technologies like
```

dipesh-rawat commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,345 @@
+are a set of features to enhance the security of virtual machines on AMD's server CPUs. SEV
+transparently encrypts the memory of each VM with a unique key. SEV can also calculate a
+signature of the memory contents, which can be sent to the VM's owner as an attestation that
+the inital guest memory was not manipulated.
```
```suggestion
the initial guest memory was not manipulated.
```

dipesh-rawat commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,345 @@
+but on the other hand applications get built with very minimal TCB.
+
+To avoid the need to build custom applications and to be able to move to process based confidential
+computing easier, an emerging approach is to use library OSes that help to run native unmodified Linux applications
```
```suggestion
computing easily, an emerging approach is to use library OSes that help to run native unmodified Linux applications
```

dipesh-rawat commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,345 @@
+[by](https://www.ibr.cs.tu-bs.de/users/mahhouk/papers/eurosec2021.pdf)
+[different papers](https://dl.acm.org/doi/fullHtml/10.1145/3533737.3535098). The chosen SDK/library
+OS, the application itself, as well as the resource requirements (especially large memory requirements)
+have a huge impact on performance. If an application is well suited to run inside an enclave a single-digit
```
```suggestion
have a huge impact on performance. If an application is well suited to run inside an enclave, a single-digit
```

dipesh-rawat commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,345 @@
+
+The presented Confidential Computing platforms show that both the isolation of a single container
+process and therefore minimization of the trusted computing base, and the isolation of a
+full virtual machine are possible. This already enabled a lot of interesting and secure
```
```suggestion
full virtual machine are possible. This has already enabled a lot of interesting and secure
```

dipesh-rawat commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,345 @@
+[Occlum](https://occlum.io/) and [Gramine](https://gramineproject.io/) are examples of open source
+library OS projects that can be used to run unmodified applications in SGX enclaves. They
+are member projects under the CCC but similar projects and products maintained by companies
+also exist. With these libOS projects existing containerized applications can be
```
```suggestion
also exist. With these libOS projects, existing containerized applications can be
```

dipesh-rawat commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,345 @@
+If you are currently working on a high-security product that struggles to run in the
+public cloud due to legal requirements or are looking to bring the privacy and security
+of your cloud-native project to the next level: Reach out to all the great projects
+we have highlighted! Everyone is keen to improve security of our ecosystem and you can
```
```suggestion
we have highlighted! Everyone is keen to improve the security of our ecosystem and you can
```

datosh commented:
Thanks for the great suggestions @dipesh-rawat. I accepted them all.

@sftim how many reviewers do we need on this? Would it be helpful if @m1ghtym0 brings in another reviewer?

nate-double-u commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,345 @@
+
+[Occlum](https://occlum.io/) and [Gramine](https://gramineproject.io/) are examples of open source
+library OS projects that can be used to run unmodified applications in SGX enclaves. They
+are member projects under the CCC but similar projects and products maintained by companies
```
```suggestion
are member projects under the CCC, but similar projects and products maintained by companies
```

nate-double-u commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,345 @@
+If you are currently working on a high-security product that struggles to run in the
+public cloud due to legal requirements or are looking to bring the privacy and security
+of your cloud-native project to the next level: Reach out to all the great projects
+we have highlighted! Everyone is keen to improve the security of our ecosystem and you can
```
```suggestion
we have highlighted! Everyone is keen to improve the security of our ecosystem, and you can
```

nate-double-u reviewed (commented):
Hi, thanks for this @datosh!

I've gone through and done a copy-edit pass. Some of the word choice suggestions are just that, I thought they'd make things a bit clearer, they are up to you to accept or not.

datosh commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,345 @@
+need to be designed and implemented to take the trusted/untrusted isolation boundaries into account
+but on the other hand applications get built with very minimal TCB.
+
+To avoid the need to build custom applications and to be able to move to process based confidential
+computing easily, an emerging approach is to use library OSes that help to run native unmodified Linux applications
+inside SGX enclaves. A library OS intercepts all application requests to host OS and processes them securely
+without the application knowing it's running a TEE.
```
Thanks. This reads a lot better and the meaning is unchanged!

datosh commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,345 @@
+inside SGX enclaves. A library OS intercepts all application requests to host OS and processes them securely
+without the application knowing it's running a TEE.
+
+The 3rd generation Xeon CPUs (aka Ice Lake Server - "ICX") and later did switch to using a technology called
```
```suggestion
The 3rd generation Xeon CPUs (aka Ice Lake Server - "ICX") and later generations did switch to using a technology called
```

I think we wanted to call out that it is the 3rd generation and all later generations that implement this new feature. I will repeat generation to make this clearer.

datosh commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,345 @@
+
+## Feature comparison
+
+In the following, we will dive a little bit deeper into the specific implementations,
```
```suggestion
These chapters will dive a little bit deeper into the specific implementations,
```

datosh commented:
Thanks @nate-double-u for taking the time. These were great suggestions. I accepted most and made two alternative suggestions (based on your feedback) which I accepted. 

@nate-double-u / @dipesh-rawat are you able to :+1: this PR so we can move to closer to publication or who needs to do the final approval here, @sftim ?

sftim commented:
Any blog approver can approve this, or a SIG Docs chair or tech lead. Details in https://github.com/kubernetes/website/blob/main/OWNERS_ALIASES if folks want to look.

I'm happy to go with the 6th of July as publication date. @datosh please feel free to Make It So.

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,346 @@
+---
+layout: blog
+title: "Confidential Kubernetes: Use Confidential Virtual Machines and Enclaves to improve your cluster security."
```
```suggestion
title: "Confidential Kubernetes: Use Confidential Virtual Machines and Enclaves to improve your cluster security"
```

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,346 @@
+
+## Feature comparison
+
+These chapters will dive a little bit deeper into the specific implementations,
```
```suggestion
These sections of the article will dive a little bit deeper into the specific implementations,
```

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,346 @@
+
+## Where are we today? Vendors, limitations, and FOSS landscape
+
+As we have seen in the previous sections, Confidential Computing is a powerful new concept
```
```suggestion
As we hope you have seen from the previous sections, Confidential Computing is a powerful new concept
```

sftim commented on content/en/blog/_posts/2023-01-17-confidential-kubernetes.md:
```diff
@@ -0,0 +1,346 @@
+to improve security, but we are still in the (early) adoption phase. New products are
+starting to emerge to take advantage of the unique properties.
+
+Google and Microsoft are the first major cloud providers to have confidential offerings.
```
```suggestion
Google and Microsoft are the first major cloud providers to have confidential offerings that
can run unmodified applications inside a protected boundary.
```
Nitro enclaves _are_ a thing (and have been since 2020ish), but not for unmodified workloads.

sftim reviewed (commented):
Here are some nits.

I think we could merge this as-is, so long as we fix the publication date [2 places BTW]. If we merge as-is, we could then _optionally_ tweak after that merge, accounting for the feedback in this review.

nate-double-u commented:
> I'm happy to go with the 6th of July as publication date. @datosh please feel free to Make It So.

I was going to suggest the same date, so this has an LGTM from me.

/lgtm

datosh commented:
@sftim I added your suggestions. Thanks!

If I read the code owners document correct either @sftim or @nate-double-u could you approve this PR?

sftim commented:
There's 4 blog approvers, BTW.

/lgtm
If you know how to @datosh, please squash this to one commit, keeping the same merge base. (That's tidier)

datosh commented:
> There's 4 blog approvers, BTW.

Do I need an approval of all 4? I understood the bot message that one would be sufficient.

> /lgtm If you know how to @datosh, please squash this to one commit, keeping the same merge base. (That's tidier)

Done

sftim commented:
Any of the 4 (and therefore, not just me or @nate-double-u) can approve.

sftim commented:
/hold cancel
/approve
